src/main/java/servlets/ProcessPaymentServlet.java

```diff
@@ -45,18 +46,20 @@ public void service(HttpServletRequest req, HttpServletResponse res) throws IOEx
             List<Cart> cartItems = null;
             if (session.getAttribute("cartItems") != null)
                 cartItems = (List<Cart>) session.getAttribute("cartItems");
-            for (Cart cart : cartItems) {
-                Book book = cart.getBook();
-                double bPrice = book.getPrice();
-                String bCode = book.getBarcode();
-                String bName = book.getName();
-                String bAuthor = book.getAuthor();
-                int availableQty = book.getQuantity();
-                int qtToBuy = cart.getQuantity();
-                availableQty = availableQty - qtToBuy;
-                bookService.updateBookQtyById(bCode, availableQty);
-                pw.println(this.addBookToCard(bCode, bName, bAuthor, bPrice, availableQty));
-                session.removeAttribute("qty_" + bCode);
+            if (cartItems != null) {
```

Comment:
This check shouldn't be necessary, because the call to `getAttribute` above guarantees that the item is present. Either you need to annotate `getAttribute` as `@Deterministic`, or change the test so that `getAttribute` is only called once. I would do the latter.